From bf61f35cb452bfd473285b4cf2fc72dd28bbd248 Mon Sep 17 00:00:00 2001
From: Ivan Romanov <drizt@land.ru>
Date: Thu, 29 Oct 2015 02:52:20 +0500
Subject: [PATCH 06/11] gnupg: fixed keys is not updating

On Windows gpg uses ANSI codepage to output homedir and keyrings
paths. No way to output in utf-8. Gpg returns utf-8 strings output
only when --with-colons option is used.
---
 plugins/qca-gnupg/gpgaction.cpp | 28 ++++++++++++++++++++--------
 plugins/qca-gnupg/gpgaction.h   |  1 +
 2 files changed, 21 insertions(+), 8 deletions(-)

diff --git a/plugins/qca-gnupg/gpgaction.cpp b/plugins/qca-gnupg/gpgaction.cpp
index f86f839..96729e6 100644
--- a/plugins/qca-gnupg/gpgaction.cpp
+++ b/plugins/qca-gnupg/gpgaction.cpp
@@ -220,6 +220,7 @@ GpgAction::GpgAction(QObject *parent)
 	: QObject(parent)
 	, proc(this)
 	, dtextTimer(this)
+	, utf8Output(false)
 {
 	dtextTimer.setSingleShot(true);
 
@@ -319,6 +320,7 @@ void GpgAction::start()
 		args += "--with-fingerprint";
 		args += "--with-fingerprint";
 		args += "--list-secret-keys";
+		utf8Output = true;
 		readText = true;
 		break;
 	}
@@ -329,6 +331,7 @@ void GpgAction::start()
 		args += "--with-fingerprint";
 		args += "--with-fingerprint";
 		args += "--list-public-keys";
+		utf8Output = true;
 		readText = true;
 		break;
 	}
@@ -760,14 +763,23 @@ void GpgAction::processResult(int code)
 
 	// put stdout and stderr into QStrings
 
-	// FIXME: on Windows gpg returns --with-colons in
-	// utf-8 charset but for -k or -K it uses
-	// console output charset (which may will be differs
-	// then system charse). Will be wait a resolving of
-	// QTBUG-13303 https://bugreports.qt-project.org/browse/QTBUG-13303
-	// After it need to make some changes.
-	QString outstr = QString::fromUtf8(buf_stdout);
-	QString errstr = QString::fromUtf8(buf_stderr);
+	QString outstr;
+	QString errstr;
+	
+#ifdef Q_OS_WIN
+	if (!utf8Output)
+	{
+		outstr = QString::fromLocal8Bit(buf_stdout);
+		errstr = QString::fromLocal8Bit(buf_stderr);
+	}
+	else
+	{
+#endif
+		outstr = QString::fromUtf8(buf_stdout);
+		errstr = QString::fromUtf8(buf_stderr);
+#ifdef Q_OS_WIN
+	}
+#endif
 
 	if(collectOutput)
 		appendDiagnosticText(QString("stdout: [%1]").arg(outstr));
diff --git a/plugins/qca-gnupg/gpgaction.h b/plugins/qca-gnupg/gpgaction.h
index b8fb189..d29109a 100644
--- a/plugins/qca-gnupg/gpgaction.h
+++ b/plugins/qca-gnupg/gpgaction.h
@@ -119,6 +119,7 @@ private:
 	bool need_submitPassphrase, need_cardOkay;
 	QString diagnosticText;
 	QCA::SafeTimer dtextTimer;
+	bool utf8Output;
 
 #ifdef GPG_PROFILE
 	QTime timer;
-- 
2.5.0

